mobile-browser-based-version/src/core/dataset/data_loader/image_loader.ts

```diff
@@ -0,0 +1,69 @@
+import { Dataset } from '../dataset_builder'
+import { DataLoader, Source, DataConfig } from './data_loader'
+import * as tf from '@tensorflow/tfjs'
+import fs from 'fs'
+
+/**
+ * TODO @s314cy:
+ * Load labels and correctly match them with their respective images, with the following constraints:
+ * 1. Images are given as 1 image/1 file
+ * 2. Labels are given as multiple labels/1 file, each label file can contain a different amount of labels
+ */
+
+export class ImageLoader extends DataLoader {
+  load (image: Source, config: DataConfig): Dataset {
+    let tensorContainer: tf.TensorContainer
+    if (config.labels === undefined) {
+      tensorContainer = ImageLoader.readImageFrom(image)
+    } else {
+      tensorContainer = {
+        xs: [ImageLoader.readImageFrom(image)],
+        ys: [config.labels[0]]
+      }
+    }
+    return tf.data.array([tensorContainer])
+  }
+
+  loadAll (images: Source[], config: DataConfig): Dataset {
+    if (config.labels === undefined) {
+      return tf.data.generator(() => {
+        let index = 0
+        const iterator = {
+          next: () => {
+            return {
+              value: ImageLoader.readImageFrom(images[index++]),
+              done: index >= images.length - 1
+            }
+          }
+        }
+        return iterator
+      })
+    } else {
+      return tf.data.generator(() => {
+        let index = 0
+        const iterator = {
+          next: () => {
+            return {
+              value: { xs: '', ys: '' },
```

Comment:
The if and else return object look very similar, maybe we can use the same but only change the `value` of the return?

in pseudo code:

f(images, index) = ImageLoader.readImageFrom(images[index++])

g(images, index) = { xs: '', ys: '' }

valueFun = if (config.labels === undefined) f else g

and then have valueFun(images, index) called to set the { value:  valueFun(images, index) } 

maybe if we define f and g using anonymous fun we get less code? (just an idea)

Reply:
Good idea, let's hope the scope resolves correctly 👍